javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,890 @@
+          throw new UnsupportedEncodingException(encoding).initCause(e)
+        case e: UnsupportedCharsetException =>
+          throw new UnsupportedEncodingException(encoding).initCause(e)
+      }
+
+      _String.lastCharset = charset
+    }
+    charset
+  }
+
+  def getBytes(encoding: Charset): Array[scala.Byte] = {
+    val buffer = encoding.encode(CharBuffer.wrap(value, offset, count))
+    val bytes  = new Array[scala.Byte](buffer.limit())
+    buffer.get(bytes)
+    bytes
+  }
+
+  def getChars(start: Int, end: Int, buffer: Array[Char], index: Int): Unit = {
+    if (0 <= start && start <= end && end <= count) {
+      System.arraycopy(value, start + offset, buffer, index, end - start)
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  override def hashCode(): Int = {
+    if (cashedHashCode == 0) {
+      if (count == 0) {
+        return 0
+      } else {
+        var hash = 0
+
+        var i = offset
+        while (i < count + offset) {
+          hash = value(i) + ((hash << 5) - hash)
+          i += 1
+        }
+
+        cashedHashCode = hash
+      }
+    }
+    cashedHashCode
+  }
+
+  def indexOf(c: Int, _start: Int): Int = {
+    var start = _start
+    if (start < count) {
+      if (start < 0) {
+        start = 0
+      }
+      if (c >= 0 && c <= Character.MAX_VALUE) {
+        var i = offset + start
+        while (i < offset + count) {
+          if (value(i) == c) {
+            return i - offset
+          } else {
+            i += 1
+          }
+        }
+      } else if (c > Character.MAX_VALUE && c <= Character.MAX_CODE_POINT) {
+        var i = start
+        while (i < count) {
+          val codePoint = codePointAt(i)
+          if (codePoint == c) {
+            return i
+          } else if (codePoint >= Character.MIN_SUPPLEMENTARY_CODE_POINT) {
+            i += 1
+          }
+
+          i += 1
+        }
+      }
+    }
+    -1
+  }
+
+  def indexOf(c: Int): Int = indexOf(c, 0)
+
+  def indexOf(string: _String): Int = indexOf(string, 0)
+
+  def indexOf(subString: _String, _start: Int): Int = {
+    var start = _start
+    if (start < 0) {
+      start = 0
+    }
+    val subCount = subString.count
+    if (subCount > 0) {
+      if (subCount + start > count) {
+        return -1
+      } else {
+        val target    = subString.value
+        val subOffset = subString.offset
+        val firstChar = target(subOffset)
+        val end       = subOffset + subCount
+        while (true) {
+          val i = indexOf(firstChar, start)
+          if (i == -1 || subCount + i > count) {
+            return -1
+          }
+
+          val o1 = offset + i
+          val o2 = subOffset
+          while (o2 < end && value(o1) == target(o2)) {}
+          if (o2 == end) {
+            return i
+          }
+          start = i + 1
+        }
+      }
+    }
+    if (start < count) start else count
+  }
+
+  def intern(): _String = ???
```

Comment:
I recommend commenting that out if it is not correctly implemented, so that the linker can report linking errors, rather than silently producing an incorrect program.